lib/client/weblogin_test.go

```diff
@@ -94,3 +100,112 @@ func TestPlainHttpFallback(t *testing.T) {
 		})
 	}
 }
+
+func TestSSHAgentPasswordlessLogin(t *testing.T) {
+	silenceLogger(t)
+
+	clock := clockwork.NewFakeClockAt(time.Now())
+	sa := newStandaloneTeleport(t, clock)
+	webID := sa.WebAuthnID
+	device := sa.Device
+
+	ctx := context.Background()
+
+	// Prepare client config, it won't change throughout the test.
+	cfg := client.MakeDefaultConfig()
+	cfg.AddKeysToAgent = client.AddKeysToAgentNo
+	// Replace "127.0.0.1" with "localhost". The proxy address becomes the origin
+	// for Webauthn requests, and Webauthn doesn't take IP addresses.
+	cfg.WebProxyAddr = strings.Replace(sa.ProxyWebAddr, "127.0.0.1", "localhost", 1 /* n */)
+	cfg.KeysDir = t.TempDir()
+	cfg.InsecureSkipVerify = true
+
+	// Reset functions after tests.
+	oldWebauthn := *client.PromptWebauthn
+	t.Cleanup(func() {
+		*client.PromptWebauthn = oldWebauthn
+	})
+
+	solvePwdless := func(ctx context.Context, origin string, assertion *wanlib.CredentialAssertion, prompt wancli.LoginPrompt) (*proto.MFAAuthenticateResponse, error) {
+		car, err := device.SignAssertion(origin, assertion)
+		if err != nil {
+			return nil, err
+		}
+		resp := &proto.MFAAuthenticateResponse{
+			Response: &proto.MFAAuthenticateResponse_Webauthn{
+				Webauthn: wanlib.CredentialAssertionResponseToProto(car),
+			},
+		}
+		resp.GetWebauthn().Response.UserHandle = webID
+
+		return resp, nil
+	}
+
+	tc, err := client.NewClient(cfg)
+	require.NoError(t, err)
+	key, err := client.NewKey()
+	require.NoError(t, err)
+
+	tests := []struct {
+		name                 string
+		customPromptWebauthn func(ctx context.Context, origin string, assert *wanlib.CredentialAssertion, p wancli.LoginPrompt, _ *wancli.LoginOpts) (*proto.MFAAuthenticateResponse, string, error)
+		customPromptLogin    wancli.LoginPrompt
+	}{
+		{
+			name: "with custom prompt",
+			customPromptWebauthn: func(ctx context.Context, origin string, assert *wanlib.CredentialAssertion, p wancli.LoginPrompt, _ *wancli.LoginOpts) (*proto.MFAAuthenticateResponse, string, error) {
+				_, ok := p.(*customPromptLogin)
+				require.True(t, ok)
+				resp, err := solvePwdless(ctx, origin, assert, p)
+				return resp, "", err
+			},
+			customPromptLogin: &customPromptLogin{},
```

Comment:
Could we assert that the custom prompt is indeed used? Otherwise it's hard to say, just from the test, that it does make a difference.

Reply:
i think this is what you meant: https://github.com/gravitational/teleport/pull/14759/commits/4d2a060b5825f090bb0ef2a9f3ff7d9bb858e25d

Reply:
That helps, but I meant something like this in the test body:

```go
		*client.PromptWebauthn = test.customPromptWebauthn
		_, err = client.SSHAgentPasswordlessLogin(ctx, req)
		require.NoError(t, err)

		if test.customPromptWebauthn != nil {
			require.True(t, test.customPromptWebauthn.called, "Custom prompt present but not called")
		}
```

Think of it this way: if we were to delete all custom prompt logic from the production code, this test would still pass, which means that it's missing something.

Reply:
ah ok, if i understood correctly, this part of the code could be deleted, and the test could still pass, so in order to detect that, i created a flag and turned it on to detect that this part of the code is called:

```
        mfaResp, _, err := promptWebauthn(ctx, webURL.String(), challenge.WebauthnChallenge, prompt, &wancli.LoginOpts{
		User:                    login.User,
		AuthenticatorAttachment: login.AuthenticatorAttachment,
	})
	if err != nil {
		return nil, trace.Wrap(err)
	}
```

Reply:
LGTM!